Expand PyTest and Sonar integrations

javajawa opened the pull request:
## Install pytest modules for more flexibility
    
- `pytest-asyncio` has been added for testing async code
- `pytest-xdist` added for parallelising tests
- `pytest-github-actions-annotate-failures` for automatically generating github annotations
    
The introduction of these modules changes how we run pytest. Coverage and results are now output primarily as XML alongside the terminal versions. Tracking coverage is not enabled by default to allow for fast testing with parallelisation
    
Tests are run with 'new' first (i.e. not in the result cache), and a test taking more than a second is reported as a potential issue

## Switch Sonar reporting to an active mode
    
Rather than have Sonar automatically scan the repo based on a webhook, we instead explicitly upload the state of a request into it. This method allows us to also upload information like the coverage report, and the reports from the linters that Sonar is configured to use.
    
The ToolChain code in the tools module has been updated to output all toolchain output to files in `reports/`.

javajawa committed:
Update documentation on how to run and write tests

The files tools/lint.py and tools/test.py can now be directly invoked.
The documentation sections on testing now have content.

javajawa committed:
Reorganise requirements for more readability

javajawa committed:
Install pytest modules for more flexibility

- `pytest-asyncio` has been added for testing async code
- `pytest-xdist` added for parallelising tests
- `pytest-github-actions-annotate-failures` for automatically generating github annotations

The introduction of these modules changes how we run pytest. Coverage and results
are now output primarily as XML alongside the terminal versions.
Tracking coverage is not enabled by default to allow for fast testing with parallelisation

Tests are run with 'new' first (i.e. not in the result cache), and a test taking
more than a second is reported as a potential issue

javajawa committed:
Switch Sonar reporting to an active mode

Rather than have Sonar automatically scan the repo based on a webhook,
we instead explicitly upload the state of a request into it.
This method allows us to also upload information like the coverage report,
and the reports from the linters that Sonar is configured to use.

The ToolChain code in the tools module has been updated to output all
toolchain output to files in `reports/`.

javajawa commented:
I really like that Sonar does the coverage calculation based on the lines you have changed -- so 100% of the code I have edited has code coverage, so it shows 100% coverage on the PR info

ajCameron commented on CONTRIBUTING.md:
```diff
@@ -76,10 +76,14 @@ such as links to decisions about the structure of the code and notes features th
--->
+
+Testing an application framework which primarily deals with network interactions is always a difficult problem.
+The design of mewbot using a number of components that communicate indirectly through dataclasses partially
```
Grammar reads a little weird here - "uses"?. Perhaps expand slightly to add further clarity. Suggest instead something like

"The design of mewbot uses a number of components which communicate by passing dataclasses to each other. This assists with testing by allowing us to craft inputs which can then be injected to produce verifiable outputs."

ajCameron commented on tools/lint.py:
```diff
@@ -5,9 +5,17 @@
-from tools.common import Annotation, ToolChain
+try:  # pylint: disable=duplicate-code
+    from tools.common import Annotation, ToolChain
+except ImportError:
```
Nice addition - might be an idea to have reminder at the end of the test run as well - if we're hitting this the user probably is trying to set up a dev environment and something has gone wrong - which they should probably fix.

javajawa replied:
I'll be honest I did this mostly so `./tools/lint.py` works, which I should probably do in my my config -- I remember in the windows work there was talk about essentially an environment file which could set this up properly so `tools/` is known to be the root of the `tools.` package...

exachixkitsune replied:
The thing concerning the `.pth` file is discussed in https://github.com/mewler/mewbot/pull/49

javajawa replied:
I am coming to the conclusion that this was a really neat hack, but I should just fix the documentation to always use `python -m tools.*` and not work around (my own) environmental issues

ajCameron commented on tools/test.py:
```diff
@@ -18,73 +19,68 @@
+        if not self.coverage:
+            args.append("--dist=load")  # Distribute between processes based on load
+            args.append("--numprocesses=auto")  # Run processes equal to CPU count
+            args.append("--maxprocesses=16")  # Cap at 16 processes
```
Seems a reasonable mumber, but why the cap?

javajawa replied:
I have no good reason!

javajawa replied:
...I was going to remove it in my rebase. will do so if there's any other changes.

javajawa replied:
Removed

ajCameron commented:
Set of very reasonable upgrades! Small notes to the docs - otherwise very happy to pass it as is.

exachixkitsune commented on .github/workflows/test.yml:
```diff
@@ -9,7 +9,42 @@ on:
     types:
```
Do we want to still call this test.yml or would another name like workflows.yml work?

javajawa replied:
Maaaayyybbeee.
I'm not sure what usefully encapsulates the purpose of the file.

`review`, with `Automated Review Pipeline` as the title?

javajawa replied:
Done

exachixkitsune commented on README.md:
```diff
@@ -70,8 +70,16 @@ python3 -m examples examples/trivial_socket.yaml
-
-> :warning: However, as of this commit, no tests exist.
\ No newline at end of file
+You can run the linters via the convenience script `tools/lint.py` (or, more
+correctly, `python3 -m tools.lint`).
```
If we're saying "more correctly, thing" should the more correct option be as the suggestion?
```suggestion
You can run the linters via the convenience script `python3 -m tools.lint`
(or `tools/lint.py`).
```

javajawa replied:
Merging this thought with https://github.com/mewler/mewbot/pull/79#discussion_r956585798
I wrote that import trick to make it so the direct exec method worked, as that is the least surprising, but honestly having two subtly different ways to run the code isn't great.

exachixkitsune commented on README.md:
```diff
@@ -70,8 +70,16 @@ python3 -m examples examples/trivial_socket.yaml
+This runs the auto-formatter `black`, two opinionated linting tools in `flake8`
+and `pylint`, and the `mypy` type checker in strict mode.
+
+You can run the test framework using the convenience script `tools/test.py` (or, more
+correctly, `python3 -m tools.test`).
```
```suggestion
You can run the test framework using the convenience script `python3 -m tools.test`
(or `tools/test.py`).
```

exachixkitsune commented on CONTRIBUTING.md:
```diff
@@ -76,10 +76,14 @@ such as links to decisions about the structure of the code and notes features th
+helps with this, allowing us to craft inputs and verify outputs.
+
+Tests are performed using the `pytest` testing framework, with tests located in the `tests` folder.
+Merge requests should not reduce the overall test coverage of the framework, as calculated by Sonar
```
SHOULD or MUST? Presumably exceptions must be agreed in the PR?

javajawa replied:
I actually don't know how Sonar configures the gate, so I didn't want to write something stricter than what the tests say

exachixkitsune commented on sonar-project.properties:
```diff
@@ -0,0 +1,20 @@
+
+sonar.python.version=3.9, 3.10, 3.11
+sonar.python.coverage.reportPaths=reports/coverage.xml
+sonar.python.xunit.reportPath=reports/junit.xml
```
are we still making junit.xml files?

javajawa replied:
We are! (we are not processing them for annotations in the workflow, because the new plugin does that during the test run, but we are uploading them to sonar)

exachixkitsune commented on src/mewbot/api/registry.py:
```diff
@@ -44,6 +44,7 @@ def __call__(  # type: ignore
             raise TypeError("Attempting to create a non registrable class")
 
         obj: Any = cls.__new__(cls)  # pylint: disable=no-value-for-parameter
+        # pylint: disable=attribute-defined-outside-init
```
out of curiosity, is there a reason not to fix this?
or is it likely to be in `abc.ABCMeta`?

javajawa replied:
This is due to the typing in this magical method being very weird; logically the line abvoe we know that `obj` is of type `Component`, but asserting that in any way causes different issues. This one I have suppressed for two reasons: firstly, it is the one which the error is most 'true' for (init is not called for another 10 lines or so), secondly it has only just appeared as a detectable thing.

Frankly, this level of messing with magic methods kind of falls outside of the linter's domain

javajawa replied:
@ajCameron found a possibly neater fix for the linting but which masks the issue a little more: https://github.com/mewler/mewbot/pull/83/files#diff-2581cf6becc21d8526e55750bf2c59d4c79f271279474ae15b9dbb51158aaca8R21

javajawa replied:
No longer a part of this PR

exachixkitsune reviewed (dismissed):
only minor comments

javajawa committed:
Setup concurrency options for review pipeline

Reviews are grouped by the ref (branch/tag) they are running for.
Creating a new instances of the review pipeline (either manually or by pushing more commits)
will cause any running pipeline to be aborted.
